### ISSUE TITLE
Allow external Nuget if already in path?

### ISSUE THREAD

**Issue by paschott** [RichieBzzzt/PoshSSDTBuildDeploy#38]
We have a deployment server that's locked down to prevent downloading a copy of nuget to the working path. Nuget is already in the path. Would it be possible to override the functionality that tries to download Nuget to allow for using one already in the system path?

**Comment by RichieBzzzt** [comment]
You mentioned system path but just to confirm: the path stored as a system environment variable, as opposed to a user environment variable?
![image](https://user-images.githubusercontent.com/26192003/46076837-ad2d3680-c186-11e8-892c-dbd5b312319a.png)

**Comment by paschott** [comment]
Yes, the system/user environment variable for path.  I may have some workarounds for this, but seem to be having some other issues w/ the ".targets" file now in the installed dacfx - similar to this issue: https://developercommunity.visualstudio.com/content/problem/166536/ssdt-not-present-in-visual-studio-build-tools-2017.html

So either need to use the on-board bits properly, but not finding that "Microsoft.Data.Tools.Schema.SqlTasks.targets" file, or see if the nuget exe in the system path will work to pull in a local DacFX that I can use.  Still not sure if that will work, but if it does, it would definitely be easier. Reading through the above thread, it seems that a _lot_ of people are having issues w/ trying to use an installed set of SSDT tools. 🙁

Not said often enough, but I really do appreciate the help.

-Peter Schott
________________________________

**Comment by RichieBzzzt** [comment]
Ok this is 2 separate things:
1) w/r/t nuget in environment path var - this is doable. Had a little play around earlier in PowerShell. I'd need to write a new function and test, so I reckon I can get this done, maybe tomorrow evening.
2)check your sqlproj file. Read this post

https://bzzzt.io/post/2016-03-08-how-to-compile-sqlproj-files-using-cmdline-msbuild----errors-included-/

You may need to open up your sqlproj file and do some editing: altering where it finds the .targets file. This is form a recent project. Is your database project particularly old? Older projects don't work well with the Microsoft.Data.Tools.MSBuild because the path was different.



```xml
 <PropertyGroup>
    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">11.0</VisualStudioVersion>
    <!-- Default to the v11.0 targets path if the targets file for the current VS version is not found -->
    <SSDTExists Condition="Exists('$(MSBuildExtensionsPath)\Microsoft\VisualStudio\v$(VisualStudioVersion)\SSDT\Microsoft.Data.Tools.Schema.SqlTasks.targets')">True</SSDTExists>
    <VisualStudioVersion Condition="'$(SSDTExists)' == ''">11.0</VisualStudioVersion>
  </PropertyGroup>
  <Import Condition="'$(SQLDBExtensionsRefPath)' != ''" Project="$(SQLDBExtensionsRefPath)\Microsoft.Data.Tools.Schema.SqlTasks.targets" />
  <Import Condition="'$(SQLDBExtensionsRefPath)' == ''" Project="$(MSBuildExtensionsPath)\Microsoft\VisualStudio\v$(VisualStudioVersion)\SSDT\Microsoft.Data.Tools.Schema.SqlTasks.targets" />
```

The targets file is by defauly packaged in the Microsoft.Data.Tools.MSBuild nuget package, hence why it is used when building a project - 

https://github.com/RichieBzzzt/PoshSSDTBuildDeploy/blob/master/PoshSSDTBuildDeploy/Functions/InvokeMSBuildSSDT.ps1#L38

**Comment by paschott** [comment]
I understand that it's two different things. I modified my code to not try to use nuget and instead point the $dacfxpath to wherever it was on the server already.  That resulted in the error.

This is a new project - at least within the last 6 months or so. I was playing around with Adventureworks for a demo so created a new project in VS2017 based on that.

Hoping that getting nuget to work from the path will maybe let me pull the dacfx locally so I can just use the bits there instead of hoping that everything lines up.  I'd really like to avoid modifying each project and then every new project to try to point to the proper .targets file. 🙁  (sent a note to Andy Yu to that effect and hoping it can be passed up the chain to whoever's in charge of the dacfx installs - different paths depending on VS edition is a killer)

Thanks again.
-Peter
________________________________

**Comment by RichieBzzzt** [comment]
No worries. Once the new function is in you won't need to download nuget from the web.

BTW, how will you download msbuild.data.tools if you have no internet connection? Or is it just no connection to download nuget?

**Comment by paschott** [comment]
Nuget is already installed so I get an error trying to install it again. Then when it tries to run Nuget from the local path, the process fails because it couldn't download it.  I have internet access - just not to install nuget.
________________________________

**Comment by RichieBzzzt** [comment]
OK gotcha no worries.

**Comment by RichieBzzzt** [comment]
Ok I’ve made a start and checked in a branch with a quick change to allow path to nuget to be added. But now my daughter’s sick and my tests are failing so I won’t get to finish tonight!

2nd phase is to add a switch to check user/system path, but am worried that might end up a bit flakey. Time will tell......

Hopefully over the weekend I can steal some time to complete.

**Comment by paschott** [comment]
I appreciate the time, but please take care of yourself and your family. That's far more important than this.

I have to pause on this for some other higher priority tasks for the moment, but hope to resume a little more next week.  It's already gotten a much better/easier build than I was using a couple of years ago so I'm grateful for that.

-Peter
________________________________

**Comment by RichieBzzzt** [comment]
OK ability to specify nuget location added and merged in and latest available on GitHub/NuGet. PowerShell Gallery is proving impossible to update, due to it whinging about something.

Now available on PSGallery.

**Comment by paschott** [comment]
Thanks. Looks like the latest version is there, but our target paths have spaces in them so I think the nuget exe is complaining about the spaces.  The "output directory" probably needs to handled double quotes around that path unless I'm missing it somewhere.

Other than that it seems to work. 😊

-Peter
________________________________

**Comment by RichieBzzzt** [comment]
Closed as of #40
